engine/storage/datamotion/src/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategy.java

```diff
@@ -255,99 +448,153 @@ private Void handleCreateVolumeFromSnapshotBothOnStorageSystem(SnapshotInfo snap
 
-        result.setResult(errMsg);
+        try {
+            snapshotInfo.getDataStore().getDriver().createAsync(snapshotInfo.getDataStore(), snapshotInfo, null);
+        }
+        finally {
+            _snapshotDetailsDao.remove(snapshotDetails.getId());
+        }
+    }
 
-        callback.complete(result);
+    private SnapshotDetailsVO handleSnapshotDetails(long csSnapshotId, String name, String value) {
+        _snapshotDetailsDao.removeDetail(csSnapshotId, name);
 
-        return null;
+        SnapshotDetailsVO snapshotDetails = new SnapshotDetailsVO(csSnapshotId, name, value, false);
+
+        return _snapshotDetailsDao.persist(snapshotDetails);
     }
 
-    private Map<String, String> getSnapshotDetails(StoragePoolVO storagePoolVO, SnapshotInfo snapshotInfo) {
-        Map<String, String> details = new HashMap<String, String>();
+    private boolean canStorageSystemCreateVolumeFromVolume(SnapshotInfo snapshotInfo) {
+        boolean supportsCloningVolumeFromVolume = false;
 
-        details.put(DiskTO.STORAGE_HOST, storagePoolVO.getHostAddress());
-        details.put(DiskTO.STORAGE_PORT, String.valueOf(storagePoolVO.getPort()));
+        DataStore dataStore = _dataStoreMgr.getDataStore(snapshotInfo.getDataStore().getId(), DataStoreRole.Primary);
 
-        long snapshotId = snapshotInfo.getId();
+        Map<String, String> mapCapabilities = dataStore.getDriver().getCapabilities();
 
-        details.put(DiskTO.IQN, getProperty(snapshotId, DiskTO.IQN));
+        if (mapCapabilities != null) {
+            String value = mapCapabilities.get(DataStoreCapabilities.CAN_CREATE_VOLUME_FROM_VOLUME.toString());
 
-        details.put(DiskTO.CHAP_INITIATOR_USERNAME, getProperty(snapshotId, DiskTO.CHAP_INITIATOR_USERNAME));
-        details.put(DiskTO.CHAP_INITIATOR_SECRET, getProperty(snapshotId, DiskTO.CHAP_INITIATOR_SECRET));
-        details.put(DiskTO.CHAP_TARGET_USERNAME, getProperty(snapshotId, DiskTO.CHAP_TARGET_USERNAME));
-        details.put(DiskTO.CHAP_TARGET_SECRET, getProperty(snapshotId, DiskTO.CHAP_TARGET_SECRET));
+            supportsCloningVolumeFromVolume = new Boolean(value);
+        }
 
-        return details;
+        return supportsCloningVolumeFromVolume;
+    }
+
+    private boolean computeClusterSupportsResign(long clusterId) {
+        List<HostVO> hosts = _hostDao.findByClusterId(clusterId);
+
+        if (hosts == null) {
+            return false;
+        }
+
+        for (HostVO host : hosts) {
+            if (host == null) {
+                return false;
+            }
+
+            DetailVO hostDetail = _hostDetailsDao.findDetail(host.getId(), "supportsResign");
+
+            if (hostDetail == null) {
+                return false;
+            }
+
+            String value = hostDetail.getValue();
+
+            Boolean booleanValue = Boolean.valueOf(value);
+
+            if (booleanValue == false) {
+                return false;
+            }
+        }
```

Comment:
Are you thinking a new SQL statement like this? By the way, it looks like the editor pulls out the star that is inside of the parenthesis of my count functions.

Select count(*) = (Select count(*) From cloud.host Where cluster_id = 1) 
From cloud.host_details 
Where name = "supportsResign" 
  and value = "true" 
  and host_id in (Select id From cloud.host Where cluster_id = 1)